src/ports/postgres/modules/knn/knn.py_in

```diff
@@ -124,16 +134,171 @@ def knn_validate_src(schema_madlib, point_source, point_column_name, point_id,
 
     return k
-# ------------------------------------------------------------------------------
+# -----------------------------------------------------------------------------
+
+
+def validate_kd_tree(source_table, output_table, point_column_name, depth):
+
+    input_tbl_valid(source_table, 'kd_tree')
+    output_tbl_valid(output_table, 'kd_tree')
+    output_tbl_valid(output_table+"_tree", 'kd_tree')
+
+    _assert(is_var_valid(source_table, point_column_name),
+            "kd_tree error: {0} is an invalid column name or expression "
+            "for point_column_name param".format(point_column_name))
+    point_col_type = get_expr_type(point_column_name, source_table)
+    _assert(is_valid_psql_type(point_col_type, NUMERIC | ONLY_ARRAY),
+            "kNN Error: Feature column or expression '{0}' in train table is not"
+            " an array.".format(point_column_name))
+    if depth <= 0:
+        plpy.error("kNN Error: depth={0} is an invalid value, must be greater "
+                   "than 0.".format(depth))
+# -----------------------------------------------------------------------------
+
+
+def kd_tree(schema_madlib, source_table, output_table, point_column_name, depth, **kwargs):
+
+    with MinWarning("error"):
+        validate_kd_tree(source_table, output_table, point_column_name, depth)
+        n_features = num_features(source_table, point_column_name)
+        clauses = [" WHERE 1=1 "]
+        cutoffs = []
+        clause_counter = 0
+        for curr_level in range(depth):
+            current_feature = depth % n_features + 1
+            for curr_leaf in range(pow(2, curr_level)):
+                clause = clauses[clause_counter]
+                # get the median for current feature for rows that fall into curr_leaf
+                cutoff_sql = """
+                    SELECT percentile_disc(0.5)
+                           WITHIN GROUP (
+                            order by {point_column_name}[{current_feature}]
+                            ) as cutoff
+                    FROM {source_table}
+                    {clause}
+                    """.format(**locals())
+                cutoff = plpy.execute(cutoff_sql)[0]['cutoff']
+                cutoffs.append(cutoff if cutoff is not None else "NULL")
+                clause_counter += 1
+                clauses.append(clause + "AND {point_column_name}[{current_feature}] < {cutoff} ".
+                               format(**locals()))
+                clauses.append(clause + "AND {point_column_name}[{current_feature}] >= {cutoff} ".
+                               format(**locals()))
+
+        output_table_tree = add_postfix(output_table, "_tree")
+        plpy.execute("CREATE TABLE {0} (tree) AS "
+                     "SELECT ('{{ {1} }}')::DOUBLE PRECISION[]".
+                     format(output_table_tree,
+                            " ,".join(map(str, cutoffs))))
+
+        case_when_clause = []
```

Comment:
Assuming we remove the `WHERE` out of the clauses, this can be built in a list comprehension: 
```
case_when_clause = ["WHEN {0} THEN {1}::INTEGER".format(cond, i) 
                                     for i, cond in enumerate(clauses[-n_leaves:])]
```